mevcommitclient/mev_commit_client.go

```diff
@@ -0,0 +1,187 @@
+		return nil, fmt.Errorf("failed to connect to MEV-Commit EVM: %w", err)
+	}
+
+	validatorOptInRouter, err := validatoroptinrouter.NewValidatoroptinrouterCaller(validatorRouterAddress, l1Client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create ValidatorOptInRouter caller: %w", err)
+	}
+
+	builderRegistryCaller, err := providerRegistry.NewProviderregistryCaller(ProviderRegistryAddress, mevCommitClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create BuilderRegistry caller: %w", err)
+	}
+
+	builderRegistryFilterer, err := providerRegistry.NewProviderregistryFilterer(ProviderRegistryAddress, mevCommitClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create BuilderRegistry filterer: %w", err)
+	}
+
+	contractAbi, err := abi.JSON(strings.NewReader(abiJSON))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse contract ABI: %w", err)
+	}
+
+	return &MevCommitClient{
+		L1Address:                  l1MainnetUrl,
+		MevCommitAddress:           mevCommitUrl,
+		ValidatorRouterAddress:     validatorRouterAddress,
+		ProviderRegistryAddress:    ProviderRegistryAddress,
+		validatorOptInRouterCaller: validatorOptInRouter,
+		builderRegistryCaller:      builderRegistryCaller,
+		builderRegistryFilterer:    builderRegistryFilterer,
+		l1Client:                   l1Client,
+		mevCommitClient:            mevCommitClient,
+		contractAbi:                contractAbi,
+	}, nil
+}
+
+func (m *MevCommitClient) GetOptInStatusForValidators(pubkeys [][]byte) ([]bool, error) {
+	// Get the finalized block number
+	var finalizedBlockNumber *big.Int
+	err := m.l1Client.Client().Call(&finalizedBlockNumber, "eth_getBlockByNumber", "finalized", false)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get finalized block: %w", err)
+	}
+
+	opts := &bind.CallOpts{
+		BlockNumber: finalizedBlockNumber,
+	}
+
+	return m.validatorOptInRouterCaller.AreValidatorsOptedIn(opts, pubkeys)
+}
+
+func (m *MevCommitClient) GetActiveBuilders() ([]MevCommitProvider, error) {
+	opts := &bind.FilterOpts{
+		Start: 0,
+		End:   nil, // Latest block
+	}
+
+	iterator, err := m.builderRegistryFilterer.FilterProviderRegistered(opts, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to filter ProviderRegistered events: %w", err)
+	}
+	defer iterator.Close()
+
+	activeBuilders := make([]MevCommitProvider, 0)
+	for iterator.Next() {
+		isValid, err := m.IsBuilderValid(iterator.Event.Provider)
+		if err != nil {
+			return nil, fmt.Errorf("failed to check if builder is valid: %w", err)
+		}
+		if isValid {
+			activeBuilders = append(activeBuilders, MevCommitProvider{
+				Pubkey:     iterator.Event.BlsPublicKey,
+				EOAAddress: iterator.Event.Provider,
+			})
+		}
+	}
+
+	if err := iterator.Error(); err != nil {
+		return nil, fmt.Errorf("error iterating over ProviderRegistered events: %w", err)
+	}
+
+	return activeBuilders, nil
+}
+func (m *MevCommitClient) ListenForActiveBuildersEvents() (<-chan MevCommitProvider, error) {
```

Comment:
It might be better to have 2 goroutines here, 1 to listen for registrations and other to listen for funds slashed and then updating it to redis here itself instead of returning a channel. Then the housekeeper can just call a function here to check.

Reply:
Sure for now, ideally, we should emit a provider de-registered event as well.

Reply:
I'll update the contracts and do simplifications after, but I'll go with the approach you're suggesting for now.